### PR TITLE
Fix bugs in TableAggregateByKey and MatrixAggregateRowsByKey

### LINKED ISSUE #3721: group_cols_by on a key field fails
A MT with 2 column keys, using one of them to group_cols_by doesn't seem to work:
```
Traceback (most recent call last):
  File "/tmp/750f9b1ef7f342459ff8104a096bfb64/pyscripts_Up_AIf.zip/gnomad_hail/utils/slack.py", line 95, in try_slack
    func(*args)
  File "/tmp/750f9b1ef7f342459ff8104a096bfb64/assign_subpops.py", line 110, in main
    pca_platforms_mt = pca_mt.group_cols_by(pca_mt.data_type).aggregate(call_rate = hl.agg.fraction(hl.is_defined(pca_mt.GT)))
  File "/tmp/750f9b1ef7f342459ff8104a096bfb64/hail-devel-3b96b54a1fa7.zip/hail/matrixtable.py", line 325, in aggregate
    base, rename = self._process_keys(base)
  File "/tmp/750f9b1ef7f342459ff8104a096bfb64/hail-devel-3b96b54a1fa7.zip/hail/matrixtable.py", line 64, in _process_keys
    left = left.annotate_cols(**new_col_fields)
  File "/tmp/750f9b1ef7f342459ff8104a096bfb64/hail-devel-3b96b54a1fa7.zip/hail/matrixtable.py", line 950, in annotate_cols
    e = get_annotate_exprs(caller, named_exprs, self._col_indices)
  File "/tmp/750f9b1ef7f342459ff8104a096bfb64/hail-devel-3b96b54a1fa7.zip/hail/utils/misc.py", line 339, in get_annotate_exprs
    check_keys(k, indices)
  File "/tmp/750f9b1ef7f342459ff8104a096bfb64/hail-devel-3b96b54a1fa7.zip/hail/utils/misc.py", line 310, in check_keys
    raise ExpressionException(msg)
hail.expr.expressions.base_expression.ExpressionException: cannot overwrite key field 'data_type' with annotate, select or drop; use key_by to modify keys.
```

### LINKED ISSUE #3787: TableAggregateByKey has an infinite loop
if `hasNext` is called while `current != null` and the `keyOrd.equiv` doesn't evaluate to true (I think this will happen if you have a unique key) then you can call `next` infinitely, returning the same element.

### PR DESCRIPTION
**Pull request by tpoterba**
Fixes #3787
fixes #3721

### CODE REVIEW

File: python/hail/matrixtable.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ def _process_keys(self, left):
                     f = self._parent._fields_inverse[v]
                 else:
                     f = Env.get_uid()
-                new_col_fields[f] = v
+                    new_col_fields[f] = v
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: I hit problems in the tests here. We don't need to annotate a field if it already exists.

Comment (tpoterba) [reply python/hail/matrixtable.py]: actually this fixes  #3721 too

**Comment by tpoterba** [comment]
fixes #3721

**Comment by danking** [comment]
the fixes clauses have to be in the main PR description, I added it there.

File: src/main/scala/is/hail/expr/Relational.scala [hunk 2]
```diff
@@ -2441,30 +2439,28 @@ case class TableAggregateByKey(child: TableIR, expr: IR) extends TableIR {
...
 
-          if (rvAggs.nonEmpty) {
-            rvAggs.foreach(_.clear())
+          rvAggs.foreach(_.clear())
+
+          val region = current.region
+          rvb.set(region)
+          rvb.start(globalsType)
+          rvb.addRegionValue(globalsType, partRegion, partGlobalsOff)
+          val globals = rvb.end()
```
Comment (danking) [inline src/main/scala/is/hail/expr/Relational.scala]: oh weird, these should have been removed. I must have missed them in the rebase. I'll get them in another PR.

**Pull request merged**